CARDS-2264: Speed up the quick search

sdumitriu opened the pull request:
This changes how quick search in forms is done, now it only looks in text answers.

Before: a few seconds matching subjects with 1.5k subjects, half a minute matching forms with 3k forms.

Now: 0.1 seconds matching both forms and subjects.

To test:

- build with `mvn clean install -Pdocker`
- in `compose-cluster` run `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --composum --adminer`
- in `compose-cluster/mssql` run `python3 generate_test_YE_sql.py -n 1000 prems_sample.sql`
- in `compose-cluster` start with `docker-compose build && docker-compose up`
- at `http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import=` import the generated `prems_sample.sql` file
- open `http://localhost:8080/Subjects.importClarity` to start the import, wait 10 minutes for the import to finish
- enable quick searching in questionnaires as well, in Administration / Quick Search Widget 
- use quicksearch to check that it returns results in under one second
- make sure the "all results" page works, including switching pages and page sizes, and going to the last page
- stop docker-compose and run `docker-compose rm -vf && ./cleanup.sh` to clean up

sdumitriu committed:
CARDS-2264: Speed up the quick search

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormsQuickSearchEngine.java:
```diff
@@ -109,17 +110,18 @@ public void quickSearch(final SearchParameters query, final ResourceResolver res
         }
     }
 
-    private String getXPathQuery(final String textQuery)
```
Why had we decided to go with XPath originally?

sdumitriu replied:
I'm not sure, probably no good reason.

sdumitriu replied:
And also, xpath queries are not natively supported, they get translated to sql internally. I actually adapted the new query from what the old one was translated into.

sashaandjic reviewed (approved):
tested - the search seems super fast
In the dev branch after importing the same file, doing the search for e.g. `2266` and `4140` was a bit slower. However when  doing search for `226` or `414` it seems to get stuck searching...
![image](https://github.com/data-team-uhn/cards/assets/23243244/2836ba64-c822-4a75-a052-0642721437e8)

marta- commented:
> In the dev branch after importing the same file, doing the search for e.g. `2266` and `4140` was a bit slower. However when doing search for `226` or `414` it seems to get stuck searching...

If that's `dev` before merging #1534 , it's expeced.

sashaandjic commented:
> > In the dev branch after importing the same file, doing the search for e.g. `2266` and `4140` was a bit slower. However when doing search for `226` or `414` it seems to get stuck searching...
> 
> If that's `dev` before merging #1534 , it's expected.

Yes, it is before the #1524 was merged. I will test it once more with #1524 changes in.

sdumitriu committed:
CARDS-2264: Speed up the quick search
More optimizations: don't generate a JSON for all the matching results, just for the ones actually in the displayed page

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/QuestionnaireQuickSearchEngine.java:
```diff
@@ -52,94 +53,111 @@ public List<String> getSupportedTypes()
-        xpathQuery.append(escapedQuery);
-        xpathQuery.append("%')]");
+        xpathQuery.append("/jcr:root/Questionnaires//*[")
+            .append("(@jcr:primaryType = 'cards:Questionnaire' or @jcr:primaryType = 'cards:Question') and ")
```
Changed behavior: before matches in answer options were also supported, but I don't think that's a useful feature. Should I bring that back @marta- ?

sashaandjic reviewed (approved):
re-tested - ALL GOOD!
